api/src/main/java/io/minio/MinioClient.java

```diff
@@ -2294,6 +2297,200 @@ public void copyObject(String bucketName, String objectName, String destBucketNa
+          InvalidKeyException, NoResponseException, XmlPullParserException, ErrorResponseException,
+          InternalException, InvalidArgumentException {
+
+    if (copySources.size() == 0 || copySources.size() > MAX_MULTIPART_COUNT ) {
+      throw new InvalidArgumentException("There must be as least one and up to MAX_MULTIPART_COUNT source objects.");
+    }
+    int i = 1;
+    int totalParts = 0;
+    long totalSize = 0;
+    for (CopySource sourceObject : copySources) {
+
+      ObjectStat os = statObject(sourceObject.bucketName(), sourceObject.objectName());
+      // Error out if client side encryption is used in this source object when
+      // more than one source objects are given.
+      if ((copySources.size() > 1 )  && sourceObject.getHeaders().containsKey("x-amz-meta-x-amz-key")) {
+        throw new InvalidArgumentException("Client side encryption is used in source object"
+                + sourceObject.bucketName() + sourceObject.objectName());
+      }
+
+      // Check if a segment is specified is within object bounds
+      if ( sourceObject.length() != null && sourceObject.length() > os.length()) {
+        throw new InvalidArgumentException("Source specified length " + sourceObject.length() + " is more than  "
+                +  "size of Object " + os.length());
+      }
+
+      if ( os.length() < MIN_MULTIPART_SIZE && i < copySources.size() - 1 ) {
+        throw new InvalidArgumentException(" Source Object size " + os.length() + " is too small "
+         +  " and " + i + " it is not the last part");
+      }
+
+      totalSize = totalSize + os.length();
+      if ( totalSize > MAX_OBJECT_SIZE ) {
+        throw new InvalidArgumentException(" Cannot compose an object of size more than " + MAX_OBJECT_SIZE );
+      }
+      if ( sourceObject.length() != null ) {
+        totalParts += calculateMultipartSize(sourceObject.length())[1];
+      } else {
+        totalParts += calculateMultipartSize((int)os.length())[1];
+      }
+
+      if ( totalParts > MAX_MULTIPART_COUNT ) {
+        throw new InvalidArgumentException(" Your proposed compose object requires more than " + MAX_MULTIPART_COUNT
+         + " parts.");
+      }
+
+      //  Ensure that the object has not been changed while
+      //  we are copying data.
+      if (!sourceObject.getHeaders().containsKey("x-amz-copy-source-if-match")) {
+        sourceObject.getHeaders().put("x-amz-copy-source-if-match",os.etag());
+      }
+      long end = os.length() - 1;
+      sourceObject.getHeaders().put("x-amz-copy-source-range","bytes="  + "0-" + end);
+      i++;
+    }
+
+    // Single source object case (i.e. when only one source is
+    // involved, it is being copied wholly and at most 5GiB in
+    // size, emptyfiles are also supported).
+    if ( copySources.size() == 1  || totalSize == 0 ) {
+      CopySource  sourceObject = copySources.get(0);
+      copyObject(sourceObject.bucketName(), sourceObject.objectName(),
+              bucketName, objectName);
```

Comment:
Why aren't you using `headers()` from `sourceObject`

Reply:
Removed